src/mlx5.cpp

```diff
@@ -330,6 +318,225 @@ int gds_mlx5_get_dword_wait_info(uint32_t *ptr, uint32_t value, int flags, gds_m
 
 //-----------------------------------------------------------------------------
 
+int gds_mlx5_rollback_send(struct gds_qp *qp,
+		       struct gds_mlx5_rollback_ctx *rollback)
+{
+	int diff;
+
+	qp->bf_offset = (rollback->rollback_id & GDS_MLX5_ROLLBACK_ID_PARITY_MASK) ?
+					qp->dv_qp.bf.size : 0;
+	rollback->rollback_id &= GDS_MLX5_ROLLBACK_ID_PARITY_MASK - 1;
+
+	if (rollback->flags & GDS_MLX5_ROLLBACK_ABORT_UNCOMMITED) {
+		diff = (qp->sq_cur_post & 0xffff)
+		     - ntohl(qp->dv_qp.dbrec[MLX5_SND_DBR]);
+		if (diff < 0)
+			diff += 0x10000;
+		qp->sq_cur_post -= diff;
+	} else {
+		if (!(rollback->flags & GDS_MLX5_ROLLBACK_ABORT_LATE)) {
+			if (qp->sq_cur_post !=
+			    (rollback->rollback_id >> 32))
+				return -ERANGE;
+		}
+		qp->sq_cur_post = rollback->rollback_id & 0xffffffff;
+	}
+	return 0;
+}
+
+//-----------------------------------------------------------------------------
+
+static inline int set_datagram_seg(struct mlx5_wqe_datagram_seg *seg, gds_send_wr *wr)
+{
+        int ret = 0;
+        mlx5dv_obj dv_obj;
+        mlx5dv_ah dv_ah;
+
+        memset(&dv_ah, 0, sizeof(mlx5dv_ah));
+        dv_obj.ah.in = wr->wr.ud.ah;
+        dv_obj.ah.out = &dv_ah;
+
+        ret = mlx5dv_init_obj(&dv_obj, MLX5DV_OBJ_AH);
+        if (ret || dv_ah.av == NULL) {
+                gds_err("Error %d in mlx5dv_init_obj(..., MLX5DV_OBJ_AH)\n", ret);
+                return ret;
+        }
+
+        memcpy(&seg->av, dv_ah.av, sizeof(struct mlx5_wqe_av));
+        seg->av.dqp_dct = htobe32(wr->wr.ud.remote_qpn | MLX5_EXTENDED_UD_AV);
+        seg->av.key.qkey.qkey = htobe32(wr->wr.ud.remote_qkey);
+        return ret;
+}
+
+//-----------------------------------------------------------------------------
+
+int gds_mlx5_post_send(struct gds_qp *qp, gds_send_wr *p_ewr, gds_send_wr **bad_ewr, gds_mlx5_peer_commit *commit)
+{
+        int ret = 0;
+        unsigned int idx;
+        int size;
+        void *seg;
+        void *qend;
+        int i;
+        int nreq;
+
+        struct mlx5_wqe_ctrl_seg *ctrl_seg;
+        struct mlx5_wqe_data_seg *data_seg;
+
+        struct gds_mlx5_peer_op_wr *wr;
+
+        if (commit->entries < 3) {
+                gds_err("not enough entries in gds_mlx5_peer_commit.\n");
+                ret = EINVAL;
+                goto out;
+        }
+
+        if (p_ewr->opcode != IBV_WR_SEND) {
```

Comment:
can you file an issue about this?
for comm, we need RDMA_WRITE as well

Reply:
Filed an issue.